### PR TITLE
Mew interaction feature

### PR DESCRIPTION
**Pull request by weswalla**
# overview
Adding the ability to like/unlike, share, and reply to mews.

# technical details
The `create_mew` zome function is used for all mew types, and internally handles what to do depending on the mew type (the main difference between the types being link creation).

Currently the input object to `create_mew` is `CreateMewInput`:
```rust
#[derive(Debug, Serialize, Deserialize, SerializedBytes)]
#[serde(rename_all = "camelCase")]
enum MewTypeInput {
    Original,
    Reply(EntryHashB64),
    ReMew(EntryHashB64),
    MewMew(EntryHashB64), // QuoteTweet
}

#[derive(Debug, Serialize, Deserialize, SerializedBytes)]
#[serde(rename_all = "camelCase")]
pub struct CreateMewInput {
  mew_type: MewTypeInput,
  mew: Option<String>,
}
```
In javascript this looks like the following for each Mew type:
```javascript
const mewContents = "My Mew with #hashtag #سعيدة #😃😃😃 and $cashtag";
let createMewInput = {
  mewType: {
    original: null
  },
  mew: mewContents
}

let createReplyMewInput = {
  mewType: {
    reply: originalEntryHash
  },
  mew: "reply mew to original mew!"
}

let createReMewInput = {
  mewType: {
    reMew: originalEntryHash
  },
  mew: null
}

let createMewMewInput = {
  mewType: {
    mewMew: originalEntryHash
  },
  mew: "mewmew of original mew!"
}
```
There is also a zome function `get_feed_mew_and_context` which returns the following struct:
```rust
pub struct FeedMewWithContext {
    pub feed_mew: FeedMew,
    pub comments: Vec<EntryHashB64>,
    pub shares: Vec<EntryHashB64>,
    pub likes: Vec<AgentPubKeyB64>,
}
```
The idea with this struct is so that when rendering the MewsFeed, there also contains the information to display like/comment/share stats, as well as the necessary hashes for opening a mew thread. So the FeedMew component could also include the corresponding buttons and numbers.

## entry hash and header hash compatibility
Since only entry hashes can be added to links, I figured it would be helpful to be able to get a mew based on its entry hash in addition to its header hash. Both `get_mew` and `get_feed_mew_and_context` work with either an entry or header hash. The main reason I worked with entry hashes (like in the `FeedMewWithContext` object, as well as `MewTypeInput`) was to reduce dht ops. If we used the HeaderHash, then every time we want to link to an entry, we would first need to use the header hash to get the element then get the entry hash from that. 

Another reason for working with entry hashes is because the entry hash is readily available in the `FeedMew` struct. So when interacting with a mew (like, comment, share) the entry hash of the corresponding mew can be obtained inside the UI (`feedMew.header.entry_hash()`) and passed in as an argument to the interaction functions.

What are other peoples thoughts on this?

# UI integration
@jost-s since the `create_mew` zome function now expects a `CreateMewInput` I'm certain there are a few changes required to the UI. I haven't tried to make these changes yet, but I may have some questions when I get around to it!

### CODE REVIEW

**Comment by jost-s** [comment]
Hi @weswalla, sweet as! I'm here for any questions and also happy to make required changes to the UI. As you prefer 🙂

**Comment by weswalla** [comment]
@jost-s cool! I'm currently in the process of trying to understand the UI and identify where changes would need to be made (may take a while since I'm completely new to vue). Feel free to go ahead and make the changes if you'd like, I'm sure I could learn a lot just by looking at your commits. Alternatively, if you're willing it would be great to sync up and get oriented with the front end code!

**Comment by jost-s** [comment]
@weswalla It seems like you're finding your way into the UI. I'm available for a sync, of course.

**Comment by weswalla** [comment]
I seems like everything is working good! Lots of ui cleanup to be done and other things like viewing threads as a separate view, but those might be better as separate PRs.
@jost-s cool, let's coordinate a sync if we think it would be helpful for merging/feedback/specing out next steps :)

File: ui/src/components/FeedMewWrapper.vue [hunk 1]
```diff
@@ -29,7 +29,7 @@
   <q-item-section>
     <div>
       <q-btn
-        @click="likeMew"
+        @click="toggleLickMew"
       >
         like ({{ mew.likes.length }})
```
Comment (bierlingm) [inline ui/src/components/FeedMewWrapper.vue:35]: I would suggest changing these instances (and all others) of like to lick so we have consistency

File: ui/src/components/FeedMewWrapper.vue [hunk 2]
```diff
@@ -92,4 +98,27 @@ const hideProfile = (index: number) => {
...
+   else {
+      // lick
+      await lickMew(props.mew.mewEntryHash);
+      // refreshMew();
+   }
+};
+
+const isLickedByMe = () => {
+  // check if my agent key is in the returned list of likes
+  return props.mew.likes.includes(myAgentPubKey);
```
Comment (bierlingm) [inline ui/src/components/FeedMewWrapper.vue:155]: likes -> licks

File: ui/src/types/types.ts [hunk 3]
```diff
@@ -58,6 +58,7 @@ export interface FeedMew {
 
 export interface FeedMewWithContext {
     feedMew: FeedMew,
+    mewEntryHash: EntryHashB64,
     comments: Array<string>,
     shares: Array<string>,
     likes: Array<string>
```
Comment (bierlingm) [inline ui/src/types/types.ts:64]: likes -> licks

File: ui/src/components/FeedMewWrapper.vue [hunk 4]
```diff
@@ -92,4 +98,27 @@ const hideProfile = (index: number) => {
...
+      // refreshMew();
+   } 
+   else {
+      // lick
+      await lickMew(props.mew.mewEntryHash);
+      // refreshMew();
+   }
+};
+
+const isLickedByMe = () => {
```
Comment (bierlingm) [inline ui/src/components/FeedMewWrapper.vue:153]: Is there possibly a way to consult our own source chain for whether a particular Mew is licked by us? Might be faster and reduce dht ops?

File: ui/src/components/FeedMewWrapper.vue [hunk 5]
```diff
@@ -92,4 +98,27 @@ const hideProfile = (index: number) => {
...
+  // check if my agent key is in the returned list of likes
+  return props.mew.likes.includes(myAgentPubKey);
+};
+
+// const refreshMew = async () => {
```
Comment (bierlingm) [inline ui/src/components/FeedMewWrapper.vue:188]: Why is this commented out?

**Review by bierlingm — COMMENTED** [review]
Looks good to me, but can't pronounce a judgment of the technical soundness

**Comment by jost-s** [comment]
@weswalla Please resolve conflicts. Also when I run `npm run start`, the app crashes with an error in the browser console for me:
```
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading '1')
    at new ProfilesStore (:3000/node_modules/.vite/@holochain-open-dev_profiles.js?v=15df3ff8:640:57)
    at :3000/src/main.ts:19:17
```
Does it happen for you too?

**Comment by weswalla** [comment]
@jost-s for resolving conflicts, should I merge develop into this branch?
and I was getting some similar profile store errors on both mac and linux, but I cloned the repo in a different directory and ran the project from there and it worked in both cases. I'm not quite sure why that worked but I'm no longer getting the issue. Perhaps @guillemcordoba has some insight into profile store issues?

File: ui/src/components/FeedMewWrapper.vue (hunk 5 above)
Comment (weswalla) [reply ui/src/components/FeedMewWrapper.vue]: I was hoping to refresh a singular mew in the mews feed after reacting to it (to update the counts) but it was more difficult than I was anticipating. So I commented out the function for now and copied the pattern of refreshing the entire feed on certain events being emitted from child components. The comments have been removed.

File: ui/src/components/FeedMewWrapper.vue (hunk 4 above)
Comment (weswalla) [reply ui/src/components/FeedMewWrapper.vue:153]: the zome function `get_full_mew_and_context` which is being used here returns a vec of agent keys that have licked the mew, so no extra dht ops are required to go through the vec and check if an agents key is in that list or not. This seems like a good enough approach for now. There is definitely a more efficient approach that would work better with mews with 1000s or more licks, but I wouldn't be the right person to implement that.

**Comment by jost-s** [comment]
@weswalla Yes, in order to resolve the conflicts, you need to merge the target branch into your branch and then push the resulting commit.

Please do that and let me know how you get on.

**Review by jost-s — APPROVED** [review]
All groovy and as stated in the convo.

I've added the UI functionality as far as I could. What's missing is displaying the connection of replies and mewmews to the original. Currently it's simply displayed as another mew without context.

**Pull request merged**